cli/running/running.go

```diff
@@ -43,58 +43,156 @@ type RunOpts struct {
 }
 
 // findAppFile searches of an application init file.
-func findAppFile(appName string, cliOpts *config.CliOpts) (string, error) {
+func collectInstances(appName string, cliOpts *config.CliOpts,
+	basePath string) ([]cmdcontext.RunningCtx, error) {
 	var err error
-	appDir := cliOpts.App.InstancesAvailable
+	var appDir string
+
+	if cliOpts.App != nil {
+		appDir = cliOpts.App.InstancesAvailable
+	}
+
 	if appDir == "" {
 		if appDir, err = os.Getwd(); err != nil {
-			return "", err
+			return nil, err
 		}
+	} else if appDir[0:1] != "/" {
+		appDir = filepath.Join(basePath, appDir)
 	}
 
 	var appPath string
+	singleApp := true
+
+	// The user can specify a specific instance from the application.
+	// Example: `tt status application:server`.
+	singleInst := false
+	singleInstName := ""
+	colonIds := strings.Index(appName, ":")
+	if colonIds != -1 {
+		singleInst = true
+		appNameTmp := appName
+		appName = appNameTmp[:colonIds]
+		singleInstName = appNameTmp[colonIds+1:]
+	}
 
 	// We considering several scenarios:
 	// 1) The application starts by `appName.lua`
 	// 2) The application starts by `appName/init.lua`
-	appAbsPath, err := util.JoinAbspath(appDir, appName+".lua")
-	if err != nil {
-		return "", err
-	}
-	dirAbsPath, err := util.JoinAbspath(appDir, appName)
-	if err != nil {
-		return "", err
-	}
+	// 3) The application starts by `dirName:appName`
+	// 4) Read application list from `appName/instances.yml`
+	luaPath := filepath.Join(appDir, appName+".lua")
+	dirPath := filepath.Join(appDir, appName)
 
 	// Check if one or both file and/or directory exist.
-	_, fileStatErr := os.Stat(appAbsPath)
-	dirInfo, dirStatErr := os.Stat(dirAbsPath)
+	_, fileStatErr := os.Stat(luaPath)
+	dirInfo, dirStatErr := os.Stat(dirPath)
 
 	if !os.IsNotExist(fileStatErr) {
 		if fileStatErr != nil {
-			return "", fileStatErr
+			return nil, fileStatErr
 		}
-		appPath = appAbsPath
+		appPath = luaPath
 	} else if dirStatErr == nil && dirInfo.IsDir() {
-		appPath = path.Join(dirAbsPath, "init.lua")
-		if _, err = os.Stat(appPath); err != nil {
-			return "", err
+		// Search for instances.yml
+		instCfgPath := path.Join(dirPath, "instances.yml")
+		if _, err = os.Stat(instCfgPath); err == nil {
```

Comment:
Can we check for != nil here? And move "main" code out of this condition to avoid indentation for huge code block?
The same comment for the line above:
`else if dirStatErr == nil && dirInfo.IsDir()`

Reply:
in this place the old logic was extended without cardinal changes